src/base/constraint.ml

```diff
@@ -70,7 +70,11 @@ end = struct
 
     let to_basic x = T x
 
-    let of_basic = function T x -> x | _ -> failwith "different constructor"
+    let of_basic = function
+      | T x ->
+          x
+      | _ ->
+          raise (Failure "different constructor")
```

Comment:
I'm just replacing `failwith s` with `raise (Failure s)`. This makes a difference because in JS, we use a custom override of the `failwith` function, to have it throw JS errors instead of strings. This is very important in a couple of circumstances where the error implies an actual failure condition. Here, the error is just used for control flow and always caught a level earlier, so we don't need a pretty error.

Reply:
If you don't want to pay the cost of the exception here, can you change the code above it to use an `val of_basic : ('v, 'f) basic -> ('v, 'f) t option` instead.